### PR TITLE
Partial admission

### LINKED ISSUE #420: Allow partial admission of jobs
**What would you like to be added**:

A mode of operation that allows partial admission of jobs, for example if parallelism is x and there is only quota available for y < x, then jobs should still be admitted. 

In this mode, the job needs to specify some minimum number of pods that is less than parallelism. To force the job controller to start only what was allowed, we need to reset parallelism to match the partial assignment, we can do that in the same update request that unsuspends the job

**Why is this needed**:

Not all jobs require all-or-nothing.

**Completion requirements**:

This enhancement requires the following artifacts:

- [X] Design doc
- [X] API change
- [X] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Adds support for partial workload admission.

#### Which issue(s) this PR fixes:

Fixes #420 

#### Does this PR introduce a user-facing change?

```release-note
Add support for partial workload admission
```

### CODE REVIEW

**Comment by tenzen-y** [comment]
I queued this PR to my tomorrow review queue.

File: pkg/controller/jobs/job/job_webhook.go [hunk 1]
```diff
@@ -103,11 +126,23 @@ func (w *JobWebhook) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.
 	return validateUpdate(&Job{oldJob}, &Job{newJob}).ToAggregate()
 }
 
-func validateUpdate(oldJob, newJob jobframework.GenericJob) field.ErrorList {
+func validateUpdate(oldJob, newJob *Job) field.ErrorList {
 	allErrs := validateCreate(newJob)
 	allErrs = append(allErrs, jobframework.ValidateUpdateForParentWorkload(oldJob, newJob)...)
+	allErrs = append(allErrs, jobframework.ValidateUpdateForOriginalPodSetsInfo(oldJob, newJob)...)
 	allErrs = append(allErrs, jobframework.ValidateUpdateForOriginalNodeSelectors(oldJob, newJob)...)
 	allErrs = append(allErrs, jobframework.ValidateUpdateForQueueName(oldJob, newJob)...)
```
Comment (alculquicondor) [inline pkg/controller/jobs/job/job_webhook.go:134]: these should all be inside one function: `jobframework.ValidateUpdate`?

File: pkg/controller/jobs/job/job_controller.go [hunk 2]
```diff
@@ -173,36 +178,47 @@ func (j *Job) ReclaimablePods() []kueue.ReclaimablePod {
...
 }
 
-func (j *Job) RestoreNodeAffinity(nodeSelectors []jobframework.PodSetNodeSelector) {
-	if len(nodeSelectors) == 0 || equality.Semantic.DeepEqual(j.Spec.Template.Spec.NodeSelector, nodeSelectors[0].NodeSelector) {
+func (j *Job) RestorePodSetsInfo(podSetsInfo []jobframework.PodSetInfo) {
+	if len(podSetsInfo) == 0 {
+		return
+	}
+
+	// if partial admission is enabled
```
Comment (alculquicondor) [inline pkg/controller/jobs/job/job_controller.go:210]: it might be better to have proper feature gates.

File: pkg/controller/jobs/job/job_controller.go [hunk 3]
```diff
@@ -237,7 +253,16 @@ func (j *Job) EquivalentToWorkload(wl kueue.Workload) bool {
 		return false
```
Comment (alculquicondor) [inline pkg/controller/jobs/job/job_controller.go:253]: I think the job framework should be providing some of the functionality here, potentially using `j.PodSets()`.

Could you explore that in a separate PR?

File: pkg/controller/jobs/job/job_controller.go [hunk 4]
```diff
@@ -237,7 +253,16 @@ func (j *Job) EquivalentToWorkload(wl kueue.Workload) bool {
...
 	}
 
-	if *j.Spec.Parallelism != wl.Spec.PodSets[0].Count {
+	ps0 := &wl.Spec.PodSets[0]
+	if mpc := j.minPodsCount(); mpc != nil {
```
Comment (alculquicondor) [inline pkg/controller/jobs/job/job_controller.go:257]: This is a bit hard to follow.
Can you extract it into a separate function `compareNullable`? You could probably implement it using generics.

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 5]
```diff
@@ -227,14 +227,40 @@ type FlavorAssignment struct {
 // be assigned immediately. Each assigned flavor is accompanied with a
 // FlavorAssignmentMode.
 func AssignFlavors(log logr.Logger, wl *workload.Info, resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, cq *cache.ClusterQueue) Assignment {
+	fullAssignment := assignFlavors(log, wl.TotalRequests, wl.Obj.Spec.PodSets, resourceFlavors, cq)
+	if fullAssignment.RepresentativeMode() == NoFit {
```
Comment (alculquicondor) [inline pkg/scheduler/flavorassigner/flavorassigner.go:231]: what if the RepresentativeMode is Preempt, but the preemption algorithm can't find any candidates for preemption?

We should still attempt a resize in that case.

Here is where it might be useful to do so in a next scheduling attempt, to keep the overall scheduling cycle simple.

**Review by alculquicondor — COMMENTED** [review]
I'm not sure this implementation covers all use cases. See my comment in the flavor assigner.

File: pkg/scheduler/flavorassigner/flavorassigner.go (hunk 5 above)
Comment (trasc) [reply pkg/scheduler/flavorassigner/flavorassigner.go:231]: I've missed that when changing from "partial before preemption", Still I'd prefer to keep it all in a schedule cycle add run the preemption nomination part in the flavor assignment.

File: test/integration/controller/job/job_controller_test.go [hunk 6]
```diff
@@ -1027,4 +1028,67 @@ var _ = ginkgo.Describe("Job controller interacting with scheduler", func() {
 		})
 	})
 
+	ginkgo.It("Should schedule jobs when partial admission is enabled", func() {
```
Comment (alculquicondor) [inline test/integration/controller/job/job_controller_test.go:1031]: Add another test for MPIJob

File: pkg/scheduler/scheduler.go
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:1]: unit tests please

File: pkg/scheduler/preemption/preemption.go
Comment (alculquicondor) [inline pkg/scheduler/preemption/preemption.go:1]: unit tests please

**Comment by alculquicondor** [comment]
I didn't review scheduler, preemption or flavor_assigner yet. I hope we can merge #806 first

**Comment by alculquicondor** [comment]
> the field is immutable for batch/job

It's mutable in k8s v1.27 or newer. But I don't think we can check this from kueue.

I'm fine leaving it as a follow up.

**Comment by alculquicondor** [comment]
Actually, let's consider starting this as an alpha feature. In that sense, 2 things need to happen:
- The code should be under a feature gate:
  - the Workload mutation webhook should actually drop the minCount if the feature is disabled on Create, but not on Update. Not sure if we should drop the annotations: let's not.
  - scheduler/preemption
- There should be an API comment for the new field saying that this feature is alpha and it will be disabled by default.

See for example: https://github.com/kubernetes/kubernetes/pull/107674

File: pkg/scheduler/scheduler.go
Comment (trasc) [reply pkg/scheduler/scheduler.go:1]: also since the changes are not functional, the fact that we have no regressions  should be sufficient.

File: pkg/scheduler/preemption/preemption.go
Comment (trasc) [reply pkg/scheduler/preemption/preemption.go:1]: also since the changes are not functional, the fact that we have no regressions should be sufficient.

File: pkg/scheduler/flavorassigner/podset_reducer.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+// from PodSets[*].Count to *PodSets[*].MinimumCount.
+type PodSetReducer[R any] struct {
+	podSets    []kueue.PodSet
+	fullCounts []int32
+	deltas     []int32
+	totalDelta int32
+	fits       func([]int32) (R, bool)
+}
+
+func NewPodSetReducer[R any](podSets []kueue.PodSet, fits func([]int32) (R, bool)) *PodSetReducer[R] {
```
Comment (alculquicondor) [inline pkg/scheduler/flavorassigner/podset_reducer.go:37]: why make it generic?

File: pkg/scheduler/flavorassigner/podset_reducer.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		psr.deltas[i] = d
+		psr.totalDelta += d
+	}
+	return psr
+}
+
+func scale(in []int32, out []int32, upFactor int32, downFactor int32) {
+	// this will panic if len(out) < len(in)
+	for i, v := range in {
+		out[i] = v * upFactor / downFactor
```
Comment (alculquicondor) [inline pkg/scheduler/flavorassigner/podset_reducer.go:59]: Do the multiplication using int64. Otherwise the calculation might overflow.

And add a test for such a case.

File: pkg/scheduler/flavorassigner/podset_reducer_test.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+		"partial not available": {
+			podSets: []kueue.PodSet{
+				*utiltesting.MakePodSet("ps1", 1).Obj(),
+				*utiltesting.MakePodSet("ps2", 2).SetMinimumCount(2).Obj(),
+			},
+			countLimit: 2,
+			wantFound:  false,
+			wantCount:  0,
+		},
+		"partial available": {
```
Comment (alculquicondor) [inline pkg/scheduler/flavorassigner/podset_reducer_test.go:32]: add one test where only one value has min

File: pkg/scheduler/flavorassigner/podset_reducer.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+func (psr *PodSetReducer[R]) Search() (R, bool) {
+	var lastGoodIdx int
+	var lastR R
+
+	if psr.totalDelta == 0 {
+		return lastR, false
+	}
+
+	current := make([]int32, len(psr.podSets))
+	idx := sort.Search(int(psr.totalDelta)+1, func(i int) bool {
```
Comment (alculquicondor) [inline pkg/scheduler/flavorassigner/podset_reducer.go:81]: For big numbers, I think this approach will skip some values.
That could be fine or it could be surprising to some users.

Given this uncertainty and #820, I think we should really limit the feature to one podset.

File: pkg/scheduler/flavorassigner/podset_reducer.go (hunk 7 above)
Comment (trasc) [reply pkg/scheduler/flavorassigner/podset_reducer.go:37]: easier testing

File: test/integration/controller/job/job_controller_test.go [hunk 11]
```diff
@@ -1027,4 +1028,76 @@ var _ = ginkgo.Describe("Job controller interacting with scheduler", func() {
...
+		origPartialAdmission := features.Enabled(features.PartialAdmission)
```
[the remark this answers is not in the document]
Comment (trasc) [reply test/integration/controller/job/job_controller_test.go:1032]: no,  you need a testing.TB which cannot be implemented (it has a private field) a workaround could be to pass nil but if something goes wrong we'll end up with a more or less cryptic panic instead of an error message.

The right way, in my opinion wold be to change `component-base`  so it will a  narrower interface, something like:
```diff
- func SetFeatureGateDuringTest(tb testing.TB, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
+ type errorer interface {
+    	Errorf(format string, args ...any)
+}
+ func SetFeatureGateDuringTest(tb errorer, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
```

After this we can use ```go 	ginkgo.GinkgoT() ```

Should I try my luck with a PR there?

File: test/integration/controller/job/job_controller_test.go [hunk 12]
```diff
@@ -1027,4 +1028,76 @@ var _ = ginkgo.Describe("Job controller interacting with scheduler", func() {
...
+				return createdJob.Spec.Suspend
+			}, util.Timeout, util.Interval).Should(gomega.Equal(pointer.Bool(false)))
+			gomega.Expect(*createdJob.Spec.Parallelism).To(gomega.BeEquivalentTo(2))
+
+			gomega.Expect(k8sClient.Get(ctx, wlKey, wl)).To(gomega.Succeed())
+			gomega.Expect(wl.Spec.PodSets[0].MinCount).ToNot(gomega.BeNil())
+			gomega.Expect(*wl.Spec.PodSets[0].MinCount).To(gomega.BeEquivalentTo(1))
+		})
+
+		ginkgo.By("changing the min parallelism the job should be suspended and its parallelism restored", func() {
```
[the remark this answers is not in the document]
Comment (trasc) [reply test/integration/controller/job/job_controller_test.go:1087]: the controller integration tests run without web-hooks, in a full setup this is restricted.

File: pkg/workload/workload.go [hunk 13]
```diff
@@ -54,6 +55,30 @@ type PodSetResources struct {
...
+		ret.Requests[k] = v
+	}
+
+	for k, v := range psr.Flavors {
+		ret.Flavors[k] = v
+	}
+	return ret
+}
+
+func (psr *PodSetResources) ScaleTo(newCount int32) *PodSetResources {
```
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/workload/workload.go]: done

File: pkg/controller/jobs/job/job_webhook.go [hunk 14]
```diff
@@ -87,10 +94,26 @@ func (w *JobWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) err
...
 	var allErrs field.ErrorList
 	allErrs = append(allErrs, jobframework.ValidateAnnotationAsCRDName(job, jobframework.ParentWorkloadAnnotation)...)
 	allErrs = append(allErrs, jobframework.ValidateCreateForQueueName(job)...)
+	allErrs = append(allErrs, validatePartialAdmissionCreate(job)...)
+	return allErrs
+}
+
+func validatePartialAdmissionCreate(job *Job) field.ErrorList {
+	var allErrs field.ErrorList
+	if strVal, found := job.Annotations[JobMinParallelismAnnotation]; found {
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/jobs/job/job_webhook.go:107]: ah, that is true. Thanks

File: test/integration/controller/job/job_controller_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply test/integration/controller/job/job_controller_test.go:1032]: It looks like GinkgoT returns an object compatible with TB, so you should be able to pass it.

File: pkg/features/kube_features.go [hunk 15]
```diff
@@ -42,5 +46,18 @@ func init() {
...
+}
+
+// Helper for `utilfeature.DefaultFeatureGate.Enabled()`
+func Enabled(f featuregate.Feature) bool {
+	return utilfeature.DefaultFeatureGate.Enabled(f)
+}
+
+func SetEnable(f featuregate.Feature, v bool) error {
+	return utilfeature.DefaultMutableFeatureGate.Set(fmt.Sprintf("%s=%v", f, v))
 }
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/features/kube_features.go:63]: We shouldn't need `SetEnable`, as per the other thread. For the others, we use them plain in the k/k codebase, so I prefer we stay consistent.

File: pkg/controller/jobs/job/job_controller.go [hunk 16]
```diff
@@ -271,6 +300,15 @@ func (j *Job) podsCount() int32 {
 	return podsCount
 }
 
+func (j *Job) minPodsCount() *int32 {
+	if strVal, found := j.GetAnnotations()[JobMinParallelismAnnotation]; found {
+		if iVal, err := strconv.Atoi(strVal); err == nil {
+			return pointer.Int32(int32(iVal))
+		}
+	}
```
Comment (tenzen-y) [inline pkg/controller/jobs/job/job_controller.go:308]: Can we output a log for parsing errors as `logger.V(3).Info()`? It would be helpful for debugging.

File: pkg/controller/jobs/mpijob/mpijob_controller.go [hunk 17]
```diff
@@ -123,7 +123,7 @@ func (j *MPIJob) PodSets() []kueue.PodSet {
 	return podSets
 }
 
-func (j *MPIJob) RunWithNodeAffinity(nodeSelectors []jobframework.PodSetNodeSelector) {
+func (j *MPIJob) RunWithPodSetsInfo(nodeSelectors []jobframework.PodSetInfo) {
```
Comment (tenzen-y) [inline pkg/controller/jobs/mpijob/mpijob_controller.go:126]:
```suggestion
func (j *MPIJob) RunWithPodSetsInfo(podSetsInfo []jobframework.PodSetInfo) {
```

File: pkg/controller/jobs/mpijob/mpijob_controller.go [hunk 18]
```diff
@@ -146,7 +146,7 @@ func (j *MPIJob) RunWithNodeAffinity(nodeSelectors []jobframework.PodSetNodeSele
 	}
 }
 
-func (j *MPIJob) RestoreNodeAffinity(nodeSelectors []jobframework.PodSetNodeSelector) {
+func (j *MPIJob) RestorePodSetsInfo(nodeSelectors []jobframework.PodSetInfo) {
```
Comment (tenzen-y) [inline pkg/controller/jobs/mpijob/mpijob_controller.go:149]:
```suggestion
func (j *MPIJob) RestorePodSetsInfo(podSetsInfo []jobframework.PodSetInfo) {
```

File: pkg/controller/jobs/rayjob/rayjob_controller.go [hunk 19]
```diff
@@ -162,7 +162,7 @@ func cloneSelectors(src map[string]string) map[string]string {
 	return dst
 }
 
-func (j *RayJob) RestoreNodeAffinity(nodeSelectors []jobframework.PodSetNodeSelector) {
+func (j *RayJob) RestorePodSetsInfo(nodeSelectors []jobframework.PodSetInfo) {
```
Comment (tenzen-y) [inline pkg/controller/jobs/rayjob/rayjob_controller.go:165]:
```suggestion
func (j *RayJob) RestorePodSetsInfo(podSetsInfo []jobframework.PodSetInfo) {
```

File: pkg/controller/jobs/rayjob/rayjob_controller_test.go [hunk 20]
```diff
@@ -298,7 +298,7 @@ func TestNodeSelectors(t *testing.T) {
 		Obj())
 
 	// run with Affinity should append or update the node selectors
```
Comment (tenzen-y) [inline pkg/controller/jobs/rayjob/rayjob_controller_test.go:300]: updating comments would be good.

File: pkg/controller/jobs/rayjob/rayjob_controller_test.go [hunk 21]
```diff
@@ -341,7 +341,7 @@ func TestNodeSelectors(t *testing.T) {
 	}
 
 	// restore should replace node selectors
```
Comment (tenzen-y) [inline pkg/controller/jobs/rayjob/rayjob_controller_test.go:343]: updating comments would be good.

File: pkg/util/maps/map.go [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package maps
+
+// Clone clones the input
+func Clone[K comparable, V any, S ~map[K]V](s S) S {
```
Comment (tenzen-y) [inline pkg/util/maps/map.go:20]: Maybe, we can replace this with `https://pkg.go.dev/golang.org/x/exp/maps` in the future when `maps` graduates to stable.
Can you add a comment?

File: pkg/util/maps/map.go
Comment (tenzen-y) [inline pkg/util/maps/map.go:1]: Can you add unit tests?

File: pkg/util/slices/slices.go [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package slices
+
+// ToMap create a map[K]V out of the provided slice s using key() to create the map keys and
```
Comment (tenzen-y) [inline pkg/util/slices/slices.go:19]:
```suggestion
// ToMap creates a map[K]V out of the provided slice s using key() to create the map keys and
```

File: pkg/util/slices/slices.go [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+// potential key conflicts.
+func ToMap[K comparable, V any, S ~[]E, E any](s S, key func(int) K, val func(int) V) map[K]V {
+	ret := make(map[K]V, len(s))
+	for i := range s {
+		ret[key(i)] = val(i)
+	}
+	return ret
+}
+
+// ToRefMap create a map[K]*S out of the provided slice s []S using key() to create the map keys.
```
Comment (tenzen-y) [inline pkg/util/slices/slices.go:32]:
```suggestion
// ToRefMap creates a map[K]*S out of the provided slice s []S using key() to create the map keys.
```

File: pkg/util/slices/slices.go
Comment (tenzen-y) [inline pkg/util/slices/slices.go:1]: Can you add unit tests?

File: pkg/scheduler/scheduler_test.go [hunk 25]
```diff
@@ -187,6 +188,9 @@ func TestSchedule(t *testing.T) {
 		wantInadmissibleLeft map[string]sets.Set[string]
 		// wantPreempted is the keys of the workloads that get preempted in the scheduling cycle.
 		wantPreempted sets.Set[string]
+
```
Comment (tenzen-y) [inline pkg/scheduler/scheduler_test.go:191]: Adding a comment would be helpful, `TODO: if PartialAdmission graduates to beta, we must remove this`.

File: pkg/util/maps/map.go (hunk 22 above)
Comment (trasc) [reply pkg/util/maps/map.go]: done

File: pkg/util/maps/map.go
Comment (trasc) [reply pkg/util/maps/map.go:1]: done

File: pkg/controller/jobs/rayjob/rayjob_controller_test.go (hunk 20 above)
Comment (trasc) [reply pkg/controller/jobs/rayjob/rayjob_controller_test.go]: done

File: pkg/scheduler/scheduler_test.go (hunk 25 above)
Comment (tenzen-y) [reply pkg/scheduler/scheduler_test.go:191]: It makes sense. Thanks for clarifying!

File: pkg/controller/jobframework/constants.go [hunk 26]
```diff
@@ -37,5 +37,13 @@ const (
...
 	// The content is a json marshaled slice of selectors.
+	//
+	// DEPRECATED: Use OriginalPodSetsInfoAnnotation.
 	OriginalNodeSelectorsAnnotation = "kueue.x-k8s.io/original-node-selectors"
+
+	// OriginalPodSetsInfoAnnotation is the annotation in which the original
+	// node selectors and podSet counts are recorded upon a workload admission.
+	// This information will be used to restore them when the job is suspended.
+	// The content is a json marshaled slice of PodSetInfo.
+	OriginalPodSetsInfoAnnotation = "kueue.x-k8s.io/original-pod-sets-info"
```
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/jobframework/constants.go:48]: Everything related to this is happening in the job framework, the job integration is agnostic of it, the GenericJob only uses the unmarshaled version, in `RunWithPod..` and `RestorePod...`

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/jobframework/constants.go:48]: I see. We might be able to clarify the hidden information to third-party developers by doing #777.

File: pkg/scheduler/flavorassigner/podset_reducer.go (hunk 10 above)
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/podset_reducer.go:81]: I spoke too soon: taking the max is enough for 2 pod sets, like seen above.

However, for (1, 2, 3), the sum gives you one more set option than max.

- max
<details>

| search | ps1 | ps2 | ps3 |
| - | - | - | - |
| 0 | 0 | 0 | 0 |
| 1 | 0 | 0 | 1 |
| 2 | 0 | 1 | 2 |
| 3 | 1 | 2 | 3 |

</details>

- sum
<details>

| search | ps1 | ps2 | ps3 |
| - | - | - | - |
| 0 | 0 | 0 | 0 |
| 1 | 0 | 0 | 0 |
| 2 | 0 | 0 | 1 |
| 3 | 0 | 1 | 1 |
| 4 | 0 | 1 | 2 |
| 5 | 0 | 1 | 2 |
| 6 | 1 | 2 | 3 |

</details>

Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/podset_reducer.go:81]: Still the question remains about whether this set of tries is acceptable to users. I prefer we do not commit to an algorithm if we don't have user feedback. Then, we can start for support for single podset only.

Comment (trasc) [reply pkg/scheduler/flavorassigner/podset_reducer.go:81]: I have tried it early on bu depending on the numbers,  if we go with max we can end up lousing some combinations, starting from 3 and 5 
```
max
0 + 0 = 0
0 + 1 = 1
1 + 2 = 3
1 + 3 = 4
2 + 4 = 6

sum
0 + 0 = 0
0 + 1 = 1
1 + 1 = 2
1 + 2 = 3
1 + 3 = 4
2 + 3 = 5
2 + 4 = 6
```
so we will not test (1,1) and (2,3)

since the binary search is O(log n) we could ignore the number of duplicates.

https://go.dev/play/p/6Hy734t-Cgx

File: pkg/controller/jobframework/reconciler.go [hunk 27]
```diff
@@ -519,34 +525,52 @@ func cloneNodeSelector(src map[string]string) map[string]string {
 	return ret
 }
 
-// getNodeSelectorsFromObjectAnnotation tries to retrieve a node selectors slice from the
+// getPodSetsInfoFromObjectAnnotation tries to retrieve a podSetsInfo slice from the
 // object's annotations fails if it's not found or is unable to unmarshal
-func getNodeSelectorsFromObjectAnnotation(obj client.Object) ([]PodSetNodeSelector, error) {
-	str, found := obj.GetAnnotations()[OriginalNodeSelectorsAnnotation]
+func getPodSetsInfoFromObjectAnnotation(obj client.Object, job GenericJob) ([]PodSetInfo, error) {
```
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/jobframework/reconciler.go:530]: One of the reasons behind going with the annotation was the fact that the workload can be deleted, In that PR initially , I was using the workload if existing and have the annotation as a "last hope"... but I was convinced :) during the review to jut use the annotation.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/jobframework/reconciler.go:530]: Yeah, it seemed like a good enough idea at the time for just node selectors. But I'm afraid that we keep adding things into annotations.

File: pkg/scheduler/flavorassigner/podset_reducer.go (hunk 10 above)
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/podset_reducer.go:81]: Yeah, I'm fine with duplicates, but is the set of combinations good enough?
I rather not go through that rabbit hole before we have more clarity about podset and rayjob.
We could check how Ray does autoscaling or preemption. But then we should still defer that for later and stick to one podset for this PR.

File: pkg/scheduler/flavorassigner/podset_reducer.go [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+		scale(psr.deltas, current, int32(i), psr.totalDelta)
+		sub(psr.fullCounts, current, current)
```
Comment (alculquicondor) [inline pkg/scheduler/flavorassigner/podset_reducer.go:83]: can you inline in a single for loop here?

File: pkg/scheduler/flavorassigner/podset_reducer_test.go [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+		"empty": {
+			podSets:    []kueue.PodSet{},
+			countLimit: 10,
+			wantFound:  false,
+			wantCount:  0,
+		},
+		"partial not available": {
+			podSets: []kueue.PodSet{
+				*utiltesting.MakePodSet("ps1", 1).Obj(),
+				*utiltesting.MakePodSet("ps2", 2).SetMinimumCount(2).Obj(),
```
Comment (alculquicondor) [inline pkg/scheduler/flavorassigner/podset_reducer_test.go:26]:
```suggestion
				*utiltesting.MakePodSet("ps2", 4).SetMinimumCount(2).Obj(),
```

Just so that there are options

File: pkg/scheduler/flavorassigner/podset_reducer.go (hunk 7 above)
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/podset_reducer.go:37]: I would prefer we don't write code just for tests, but the function used in the test seems simple enough.

**Review by alculquicondor — COMMENTED** [review]
Overall LGTM, but I'm looking for opportunities to simplify.
I think we might need to revert the annotation. Could you explore this possibility in a preparatory PR?

File: pkg/scheduler/flavorassigner/podset_reducer.go (hunk 10 above)
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/podset_reducer.go:81]: It seems that most people prefer we don't commit to multiple podsets yet. Can you strip it down?

File: test/integration/controller/job/job_controller_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (trasc) [reply test/integration/controller/job/job_controller_test.go:1085]: I'll open do followup for this if/when  kubernetes/kubernetes#118346 is merged.

File: pkg/controller/jobframework/reconciler.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/jobframework/reconciler.go:530]: we should not block this because of that, we should adapt with the last one that gets merged.

File: pkg/scheduler/flavorassigner/podset_reducer_test.go (hunk 29 above)
Comment (trasc) [reply pkg/scheduler/flavorassigner/podset_reducer_test.go:26]: the partial admission should not be possible in this case  (count needs to be == *minCount)

File: pkg/scheduler/flavorassigner/podset_reducer.go (hunk 28 above)
Comment (trasc) [reply pkg/scheduler/flavorassigner/podset_reducer.go]: done

File: pkg/scheduler/flavorassigner/podset_reducer.go (hunk 10 above)
Comment (trasc) [reply pkg/scheduler/flavorassigner/podset_reducer.go:76]: the webhook part is added

File: test/integration/controller/job/job_controller_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (trasc) [reply test/integration/controller/job/job_controller_test.go:1140]: the purpose is to check that the the restore is done properly, not to go around the web-hook.

I chose this because besides suspending the it will also block it's potential re-admission.

File: pkg/features/kube_features.go [hunk 30]
```diff
@@ -42,5 +46,18 @@ func init() {
...
+func SetFeatureGateDuringTest(tb testing.TB, f featuregate.Feature, value bool) func() {
+	return featuregatetesting.SetFeatureGateDuringTest(tb, utilfeature.DefaultFeatureGate, f, value)
+}
+
+// Helper for `utilfeature.DefaultFeatureGate.Enabled()`
+func Enabled(f featuregate.Feature) bool {
+	return utilfeature.DefaultFeatureGate.Enabled(f)
+}
+
+func SetEnable(f featuregate.Feature, v bool) error {
```
Comment (alculquicondor) [inline pkg/features/kube_features.go:61]: Add a comment that this is only meant to be used by integration tests.

File: pkg/scheduler/flavorassigner/podset_reducer.go [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+	var lastGoodIdx int
+	var lastR R
+
+	if psr.totalDelta == 0 {
+		return lastR, false
+	}
+
+	current := make([]int32, len(psr.podSets))
+	idx := sort.Search(int(psr.totalDelta)+1, func(i int) bool {
+		scaleAndSub(psr.fullCounts, psr.deltas, current, int32(i), psr.totalDelta)
```
Comment (alculquicondor) [inline pkg/scheduler/flavorassigner/podset_reducer.go:77]: Use a name that reflects the intent, not the operation.

```suggestion
		fillPodSetSizesForSearchIndex(current, psr.fullCounts, psr.deltas, int32(i), psr.totalDelta)
```

Also it's more common to have the output as first parameter, or even return it, like:

```
list = append(list, newItems...)
```

File: pkg/webhooks/workload_webhook.go [hunk 32]
```diff
@@ -96,8 +105,17 @@ func ValidateWorkload(obj *kueue.Workload) field.ErrorList {
...
-		allErrs = append(allErrs, validatePodSet(&obj.Spec.PodSets[i], specPath.Child("podSets").Index(i))...)
+		ps := &obj.Spec.PodSets[i]
+		allErrs = append(allErrs, validatePodSet(ps, specPath.Child("podSets").Index(i))...)
+		if pointer.Int32Deref(ps.MinCount, ps.Count) != ps.Count {
+			variableCountPosets++
+		}
+	}
+
+	if variableCountPosets > 1 {
+		allErrs = append(allErrs, field.Invalid(specPath.Child("podSets"), variableCountPosets, "at most one podSet can have variable count"))
```
Comment (alculquicondor) [inline pkg/webhooks/workload_webhook.go:118]:
```suggestion
		allErrs = append(allErrs, field.Invalid(specPath.Child("podSets"), variableCountPosets, "at most one podSet can use minCount"))
```

Also please add it to the API documentation (as part of the comment in the field).

File: pkg/webhooks/workload_webhook.go [hunk 33]
```diff
@@ -96,8 +105,17 @@ func ValidateWorkload(obj *kueue.Workload) field.ErrorList {
 	var allErrs field.ErrorList
 	specPath := field.NewPath("spec")
 
+	variableCountPosets := 0
 	for i := range obj.Spec.PodSets {
-		allErrs = append(allErrs, validatePodSet(&obj.Spec.PodSets[i], specPath.Child("podSets").Index(i))...)
+		ps := &obj.Spec.PodSets[i]
+		allErrs = append(allErrs, validatePodSet(ps, specPath.Child("podSets").Index(i))...)
+		if pointer.Int32Deref(ps.MinCount, ps.Count) != ps.Count {
```
Comment (alculquicondor) [inline pkg/webhooks/workload_webhook.go:112]: I would just check that `minCount` is used, because it's easier to document.

**Review by alculquicondor — COMMENTED** [review]
/approve
Just final nits

Please adjust the release note to indicate that this is specific to Job.

File: pkg/scheduler/flavorassigner/podset_reducer.go (hunk 10 above)
Comment (trasc) [reply pkg/scheduler/flavorassigner/podset_reducer.go:76]: we'll go deeper in the ray implementation with #783

File: pkg/scheduler/flavorassigner/podset_reducer.go (hunk 31 above)
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/podset_reducer.go]: https://twitter.com/thockin/status/1646625589225455616
